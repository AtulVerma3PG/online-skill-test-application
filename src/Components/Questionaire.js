import React, { Component } from "react";
import _ from "lodash";
import Question from "./Question";
import { logout } from "./utils/index";
import skill0Questionaire from "../Questionaire.json";
import skill1Questionaire from "../QuestionaireSkill1.json";
import skill2Questionaire from "../QuestionaireSkill2.json";

const skillLevels = ["Level 1", "Level 2", "Level 3"];

let initialState = {
  questionIndex: 0,
  isFalseLogin: false,
  quizTime: 600,
  questionCount: 0,
  IsQuizSubmitted: false,
  skillLevel: "",
  questionData: "",
  candidateDetails: [],
  candidateResponses: [],
  currentResponse: [],
  questionsAttempted: 0,
  candidateScore: 0,
  firstName: "",
  lastName: "",
  email: "",
  gender: "",
  password: "",
};
class Questionaire extends Component {
  constructor(props) {
    super(props);
    this.setInitialState();
    this.state = JSON.parse(localStorage.getItem("state"))
      ? JSON.parse(localStorage.getItem("state"))
      : initialState;
    this.updateLocalStorage();
  }

  componentDidMount() {
    const { history } = this.props;
    this.timer = setInterval(() => {
      // eslint-disable-next-line react/destructuring-assignment
      const newCount = this.state.quizTime - 1;
      this.setState({ quizTime: newCount >= 0 ? newCount : 0 });
      this.updateLocalStorage();
    }, 1000);
    this.updateLocalStorage();
    window.addEventListener("popstate", () => {
      history.push("/Questionaire");
    });
  }

  /**
   * Reset Timer on quiz over
   */
  resetTimer = () => {
    this.setState({
      quizTime: 600,
    });
    this.updateLocalStorage();
  };

  /**
   * Display next question on questionaire
   */
  nextQuestion = () => {
    const {
      currentResponse, questionIndex, questionData, candidateResponses,
    } = this.state;
    const isAnswerCorrect = JSON.stringify(currentResponse)
      == JSON.stringify(questionData[questionIndex].answer);
    const isQuestionAttempted = currentResponse.length > 0;
    candidateResponses.push({
      id: questionIndex,
      Question: questionData[questionIndex].question,
      Response: currentResponse,
      Answer: questionData[questionIndex].answer,
      IsAnswerCorrect: isAnswerCorrect,
      IsAttempted: isQuestionAttempted,
    });
    const questIndex = questionIndex + 1;
    this.clearResponse();
    const questionResponse = candidateResponses.filter((a) => a.id == questIndex);
    let response = [];
    if (questionResponse.length !== 0) {
      response = questionResponse[0].Response;
    }
    this.setState({
      questionIndex: questIndex,
      currentResponse: response,
    }, () => this.updateLocalStorage());
  };

  /**
   *Save the choice selected
   *
   * @param {string} event choice of question selected
   */
  choiceSelected = (event) => {
    const { currentResponse } = this.state;
    if (currentResponse.includes(event)) {
      currentResponse.splice(currentResponse.indexOf(event), 1);
    } else {
      currentResponse.push(event);
    }
    this.updateLocalStorage();
  };

  /**
   * Clear responses selected
   */
  clearResponse = () => {
    this.setState({
      currentResponse: [],
    });
    this.updateLocalStorage();
  };

  /**
   * Download JSON file of the Candidate Responses
   */
  downloadFile = async () => {
    const { candidateResponses } = this.state;
    const myData = candidateResponses;
    const fileName = "CandidateResponses";
    const json = JSON.stringify(myData);
    const blob = new Blob([json], { type: "application/json" });
    const href = await URL.createObjectURL(blob);
    const link = document.createElement("a");
    link.href = href;
    link.download = `${fileName}.json`;
    document.body.appendChild(link);
    link.click();
    document.body.removeChild(link);
    this.updateLocalStorage();
  };

  /**
   * Submit candidate Responses
   */
  submitTest = () => {
    const { candidateResponses } = this.state;
    const { history } = this.props;
    const correctAnswers = candidateResponses.filter(
      (c) => c.IsAnswerCorrect == true,
    ).length;
    const questionsAttempt = candidateResponses.filter(
      (c) => c.IsAttempted == true,
    ).length;
    logout();
    this.setState(
      {
        candidateScore: correctAnswers,
        questionsAttempted: questionsAttempt,
        IsQuizSubmitted: true,
      },
      () => {
        this.resetTimer();
        this.downloadFile();
        history.push({
          pathname: "/SubmitTest",
          state: { ...this.state },
        });
      },
    );
  };

  /**
   * Display Previous qestion in the questionaire
   */
  previousQuestion = () => {
    const { questionIndex, candidateResponses } = this.state;
    const questionResponse = candidateResponses.filter((a) => a.id == questionIndex - 1);
    let response = [];
    if (questionResponse.length !== 0) {
      response = questionResponse[0].Response;
      candidateResponses.splice(questionIndex - 1, 1);
    } else {
      candidateResponses.splice(questionIndex - 1, 1);
    }
    this.setState({
      questionIndex: questionIndex - 1,
      currentResponse: response,
    }, () => this.updateLocalStorage());
  };

setQuestion = (event) => {
  const { candidateResponses } = this.state;
  const questionResponse = candidateResponses.filter((a) => a.id == event - 1);
  let response = [];
  if (questionResponse.length !== 0) {
    response = questionResponse[0].Response;
    candidateResponses.splice(event - 1, 1);
  }
  this.setState({
    questionIndex: event - 1,
    currentResponse: response,
  }, () => this.updateLocalStorage());
}

setInitialState() {
  if (!JSON.parse(localStorage.getItem("state"))) {
    const userData = JSON.parse(localStorage.getItem("userDetails"));
    const level = _.get(this.props, "location.state.skillLevel");
    let questionsd = [];
    if (level === skillLevels[0]) {
      questionsd = skill0Questionaire.questions;
    } else if (level === skillLevels[1]) {
      questionsd = skill1Questionaire.questions;
    } else {
      questionsd = skill2Questionaire.questions;
    }
    initialState = {
      skillLevel: level,
      questionData: questionsd,
      questionCount: questionsd.length,
      firstName: userData.firstName,
      lastName: userData.lastName,
      password: userData.password,
      email: userData.email,
      gender: userData.gender,
      isRegistered: userData.isRegistered,
      questionIndex: 0,
      isFalseLogin: false,
      quizTime: 600,
      IsQuizSubmitted: false,
      candidateDetails: [],
      candidateResponses: [],
      currentResponse: [],
      questionsAttempted: 0,
      candidateScore: 0,
    };
  }
}

/**
   * Update Local storage on state change
   */
updateLocalStorage() {
  localStorage.setItem("state", JSON.stringify(this.state));
}

render() {
  const {
    questionIndex, questionData, questionCount, IsQuizSubmitted, quizTime, currentResponse,
  } = this.state;
  return (
    <Question
      questionId={questionIndex}
      questionText={
            questionData[questionIndex].question
          }
      options={questionData[questionIndex].options}
      questionCount={questionCount}
      onNext={this.nextQuestion}
      onPrevious={this.previousQuestion}
      onChange={this.choiceSelected}
      submitTest={this.submitTest}
      answer={currentResponse}
      setQuestion={this.setQuestion}
      clearResponse={this.clearResponse}
      isQuizSubmitted={IsQuizSubmitted}
      quizTime={quizTime}
    />

  );
}
}

Questionaire.propTypes = {
  history: Object.isRequired,
};

export default Questionaire;
