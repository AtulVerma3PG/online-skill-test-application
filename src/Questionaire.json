{
  "questions": [
    {
      "id": 0,
      "question": "What temperature does water boil at?",
      "options": [
        "50 degrees Celcius",
        "25 degrees Celcius",
        "100 degrees Celcius",
        "150 degrees Celcius"
      ],
      "answer": ["100 degrees Celcius"]
    },

    {
      "id": 1,
      "question": "Who wrote Julius Caesar, Macbeth and Hamlet?",
      "options": [
        "Wole Soyinka",
        "William Shakespeare",
        "Ngozi Chimamanda Adichie",
        "Dan Brown"
      ],
      "answer": ["William Shakespeare"]
    },

    {
      "id": 2,
      "question": "What did the crocodile swallow in Peter Pan?",
      "options": ["A Book", "A Computer", "A pair of shoes", "Alarm Clock"],
      "answer": ["Alarm Clock"]
    },

    {
      "id": 3,
      "question": "Which is the only mammal that can’t jump?",
      "options": ["Dog", "Elephant", "Goat", "Lion"],
      "answer": ["Elephant"]
    },

    {
      "id": 4,
      "question": "Who lived at 221B, Baker Street, London?",
      "options": [
        "Tony Stark",
        "Morgan Freeman",
        "Sherlock Holmes",
        "Samuel L. Jackson"
      ],
      "answer": ["Sherlock Holmes"]
    },

    {
      "id": 5,
      "question": "What colour is a panda?",
      "options": [
        "Green and Yellow",
        "Blue and Red",
        "Green and White",
        "Black and White"
      ],
      "answer": ["Black and White"]
    },

    {
      "id": 6,
      "question": "Where is the smallest bone in the human body?",
      "options": ["The Chest", "The Ear", "The Legs", "The Hands"],
      "answer": ["The Ear"]
    },

    {
      "id": 7,
      "question": "What does the roman numeral C represent?",
      "options": ["100", "10", "10,000", "1,000,000"],
      "answer": ["100"]
    },

    {
      "id": 8,
      "question": "What takes place in Hong Kong's Happy Valley?",
      "options": [
        "Chicken Wrestling",
        "Horse racing",
        "Street Racing",
        "Arm Wrestling"
      ],
      "answer": ["Horse racing"]
    },

    {
      "id": 9,
      "question": "Who painted the Mona Lisa?",
      "options": [
        "Alexander Graham Bell",
        "Sir Isaac Newton",
        "Leonardo Da Vinci",
        "Albert Einstein"
      ],
      "answer": ["Leonardo Da Vinci"]
    },

    {
      "id": 10,
      "question": "What’s the most important book in the Moslem religion?",
      "options": [
        "The Koran",
        "The Dictionary",
        "The Bible",
        "The Chemistry text Book"
      ],
      "answer": ["The Koran"]
    }
  ]
}
