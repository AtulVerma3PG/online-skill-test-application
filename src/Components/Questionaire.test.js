/* eslint-disable */
import React from "react";
import renderer from "react-test-renderer";
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Questionaire from "./Questionaire";

const state = {
  firstName: "",
  lastName: "",
  email: "",
  password: "",
  gender: "",
  isRegistered: false,
  formErrors: {
    firstName: "",
    lastName: "",
    gender: "",
    email: "",
    password: "",
  },
  firstNameValid: false,
  lastNameValid: false,
  emailValid: false,
  passwordValid: false,
  genderValid: false,
  formValid: false,
}

configure({adapter: new Adapter()});
test("First Submit Test Snapshot test", () => {
  localStorage.setItem("userDetails", JSON.stringify(state));
  const component = renderer.create(
    <Questionaire />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
  localStorage.removeItem("userDetails");
});