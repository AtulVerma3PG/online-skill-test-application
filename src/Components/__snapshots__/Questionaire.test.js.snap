// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`First Submit Test Snapshot test 1`] = `
<div
  className="container"
>
  <div
    className="Countdown-time"
  >
    Timer :-
     
    10
     
    :
     
    00
  </div>
  <h3
    id="question"
  >
    Question
     
    1
     
    :
     
    HTML stands for?
  </h3>
  <div
    id="altcontainer"
  >
    <div
      className="radio"
    >
      <div>
        <label
          className="radio-inline"
          htmlFor="{choice}"
        >
          <input
            checked={false}
            onChange={[Function]}
            type="checkbox"
            value="Hyper Text Markup Language"
          />
          Hyper Text Markup Language
        </label>
        <br />
      </div>
      <div>
        <label
          className="radio-inline"
          htmlFor="{choice}"
        >
          <input
            checked={false}
            onChange={[Function]}
            type="checkbox"
            value="High Text Markup Language"
          />
          High Text Markup Language
        </label>
        <br />
      </div>
      <div>
        <label
          className="radio-inline"
          htmlFor="{choice}"
        >
          <input
            checked={false}
            onChange={[Function]}
            type="checkbox"
            value="Hyper Tabular Markup Language"
          />
          Hyper Tabular Markup Language
        </label>
        <br />
      </div>
      <div>
        <label
          className="radio-inline"
          htmlFor="{choice}"
        >
          <input
            checked={false}
            onChange={[Function]}
            type="checkbox"
            value="None of these"
          />
          None of these
        </label>
        <br />
      </div>
    </div>
  </div>
  <button
    className="questionaireButton btn btn-info"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Next
  </button>
  <button
    className="questionaireButton btn btn-info"
    disabled={false}
    id="submitTest"
    onClick={[Function]}
    type="button"
  >
    Submit Test
  </button>
  <nav
    className="navbar navbar-expand navbar-light"
  >
    <a
      className="navbar-brand"
      href="#home"
    >
      Go To Question : 
    </a>
    <div
      className="mr-auto navbar-nav"
      onKeyDown={[Function]}
    >
      <a
        className="nav-link active"
        data-rb-event-key={null}
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="button"
      >
        1
      </a>
      <a
        className="nav-link"
        data-rb-event-key={null}
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="button"
      >
        2
      </a>
      <a
        className="nav-link"
        data-rb-event-key={null}
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="button"
      >
        3
      </a>
      <a
        className="nav-link"
        data-rb-event-key={null}
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="button"
      >
        4
      </a>
    </div>
  </nav>
</div>
`;
